File: E-Tour/src/main/java/com/etour/common/TourDates.java
package com.etour.common;

import java.time.LocalDate;

import com.etour.entities.Booking;
import com.etour.entities.Schedule;
import com.etour.entities.Tour;

public final class TourDates {
    private TourDates() {
    }

    public static LocalDate returnDate(Schedule schedule, Tour tour) {
        if (schedule == null || schedule.getStartDate() == null) {
            return null;
        }

        LocalDate departure = schedule.getStartDate();
        Integer days = tour != null ? tour.getDays() : null;

        if (days == null || days < 1) {
            return departure;
        }

        return departure.plusDays(days - 1L);
    }

    public static LocalDate returnDate(Booking booking) {
        if (booking == null) {
            return null;
        }
        return returnDate(booking.getSchedule(), booking.getTour());
    }

    public static boolean hasFinished(Booking booking, LocalDate today) {
        LocalDate end = returnDate(booking);
        return end != null && end.isBefore(today);
    }
}

File: ETour-DotNet/Common/TourDates.cs
using ETour.Api.Models;

namespace ETour.Api.Common;

public static class TourDates
{
    public static DateOnly? ReturnDate(Schedule schedule, Tour tour)
    {
        if (schedule?.StartDate is null)
        {
            return null;
        }

        var departure = schedule.StartDate.Value;
        var days = tour?.Days;

        if (days is null || days < 1)
        {
            return departure;
        }
        return departure.AddDays(days.Value - 1);
    }

    public static DateOnly? ReturnDate(Booking booking) =>
        booking is null ? null : ReturnDate(booking.Schedule, booking.Tour);

    public static bool HasFinished(Booking booking, DateOnly today)
    {
        var end = ReturnDate(booking);
        return end is not null && end < today;
    }
}
